### PR TITLE
added 3 open-source basemap tile providers: OSM, WIKIMEDIA, ESRI_IMAGERY

### LINKED ISSUE #9560: add tile providers for OpenStreetMap and ESRI Imagery
Hello,

**TL;DR**
This issue is to initiate a pull request to add the following tile providers to the Bokeh file **`./bokeh/tile_providers.py`:**
- OpenStreetMaps map tile provider
- Wikimedia / Wikipedia map tile provider (via OSM)
- ESRI Imagery with satellite imagery

**Background:**
In my search for a Google Maps replacement, the current Bokeh has tile providers are a good start but the Stamen and Carto tiles do not provide a combination of:
- ability for close zoom-in (map data not available)
- detail at high zoom (it's just blank, or not descriptive)
- satellite imagery

**Licensing:**
The [OpenStreetMap data](www.openstreetmap.org) is an excellent map-style alternative and the ESRI Imagery tiles are excellent satellite image replacements. OSM is entirely open-source. Esri has [generous open-source](https://www.esri.com/en-us/arcgis/open-vision/standards/open-source) and [open-data policies](https://www.esri.com/en-us/arcgis/open-vision/standards/open-data). Neither of these require api keys or registration, although responsible usage is, indeed, encouraged and enforced. Unusually high usage places a penalty on volunteer and donated hardware and network services.

But, for simple basemap overlays (underlays), prototyping for research purposes, or usage in other open-source projects, these two are outstanding.

**Rationale:**
Both of these tile providers, and others, are already integrated into [PyViz](http://pyviz.org),  [Holoviews](http://holoviews.org/), and [GeoViews](http://geoviews.org/), all of which rely on Bokeh as an underlying library. Rather than adding layers (and moving to Anaconda package management) to access those map tile providers, it seems Bokeh could (should) have these maps available directly.

**Example:**
The screenshot at this link shows multiple failed or poorly displaying zoomed-in options from the fine work presented at the [GeoViews page on working with Bokeh](http://geoviews.org/user_guide/Working_with_Bokeh.html):

https://discourse.bokeh.org/t/open-source-tile-provider-replacement-for-google-maps/4462

Marc Compere

### PR DESCRIPTION
**Pull request by comperem**
This pull request is to integrate 3 open-source basemap tile providers to `tile_providers.py`:
    (a) OpenStreetMap's OSM map tiles
    (b) OpenStreetMap's WIKIMEDIA map tiles and
    (c) ESRI's ESRI_IMAGERY tiles with satellite views

Unit tests were added but appear to be uncalled. For example, placing a failing garbage string anywhere in the unit tests allows the `test_tile_providers.py` to run without error.

This is true with `Test___all__ = verify_all(bt, ALL)` commented or uncommented. (ha)

Opening issue:
https://github.com/bokeh/bokeh/issues/9560
- [x] issues: fixes #9560 
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

File: tests/unit/bokeh/test_tile_providers.py [hunk 1]
```diff
@@ -115,13 +130,77 @@ def test_copies(self, name):
         p2 = bt.get_provider(getattr(bt, name))
         assert p1 is not p2
 
+@pytest.mark.unit
```
Comment (mattpap) [inline tests/unit/bokeh/test_tile_providers.py]: You need to add:
```py
@pytest.mark.parametrize('name', ['OSM''])
```
etc. as well. See other tests (e.g. line 110).

File: bokeh/tile_providers.py [hunk 2]
```diff
@@ -218,6 +241,10 @@ def get_provider(self, provider_name):
     STAMEN_TONER = Vendors.STAMEN_TONER
     STAMEN_TONER_BACKGROUND = Vendors.STAMEN_TONER_BACKGROUND
     STAMEN_TONER_LABELS = Vendors.STAMEN_TONER_LABELS
+    STAMEN_TONER_LABELS = Vendors.STAMEN_TONER_LABELS
```
Comment (bryevdv) [inline bokeh/tile_providers.py]: copy-paste error I think, this line duplicates the line right above

**Comment by bryevdv** [comment]
@comperem Thank you very much for this PR. There are just  a few small changes needed to get the tests green so we can merge. We have left some comments but please let us know if anything is confusing or if there is anything we can help with!

File: tests/unit/bokeh/test_tile_providers.py (hunk 1 above)
Comment (comperem) [reply tests/unit/bokeh/test_tile_providers.py]: ok, right - got it. I missed that - thanks. added

Comment (comperem) [reply tests/unit/bokeh/test_tile_providers.py]: > You need to add:
> 
> ```python
> @pytest.mark.parametrize('name', ['OSM'])
> ```
> 
> etc. as well. See other tests (e.g. line 110).

ok, right - got it - I missed that - thanks. added

File: bokeh/tile_providers.py (hunk 2 above)
Comment (comperem) [reply bokeh/tile_providers.py]: > copy-paste error I think, this line duplicates the line right above

yes, ditto - thanks - mistaken duplicate line - corrected

**Comment by bryevdv** [comment]
@comperem fantastic, thanks for the contribution! Since there are not otherwise example tests for this, whenever you get a chance please do one last thing: post a short script that uses all three of these new sources, and also a screenshot image of the output, as comment in the PR. Thanks!!

**Comment by comperem** [comment]
ok Bryan, that's a good idea. I'll put something together. Thanks for incorporating and helping me with the contribution.

**Comment by austintpg** [comment]
Hi @bryevdv, I would like to know if there is a date for the next Bokeh update that includes @comperem  contribution. I found it very useful and I would like to update Bokeh in my computer. Thank you.

**Comment by bryevdv** [comment]
We are aiming for the first half of Feb, though I can't make a hard promise, it's always possible things take longer than planned.

**Comment by austintpg** [comment]
That is great, thanks.

**Comment by sebhan2** [comment]
Hello, I wanted to ask if the OSM provider is available? I could not find it in the documentation of bokeh version 1.4.0 and would really like to use the extension. Thank you.

**Comment by comperem** [comment]
Hello,

I believe these are scheduled for the next release. I’m not a maintainer so
am uncertain.

The ESRI and OSM tiles work great. I’m so pleased to have an open source
alternative to google maps.

I’ve attached the *tile_providers.py* I’m using on my modified local
installation of v.1.4.0. If you have privileges to edit *tile_providers.py*
in */usr/local/lib/python3.6/dist-packages/bokeh* (or wherever yours is
located) you can probably get these working today. I'm on Ubuntu 18.04.4
with this slightly modified version of bokeh 1.4.0.

Marc

**Comment by sebhan2** [comment]
Hi Marc,
works like a charm! Thanks!
All the best,
Sebastian

**Comment by comperem** [comment]
Hello Sebastian,

This is good news. It's exciting. This open-source mapping with
openstreetmap.org and imagery from ESRI's open-source tile providers really
opens up a new dimension of map overlays with Bokeh. The Google maps
functionality has been in Bokeh for a while but I've always been
hesitant to commit to a development path that includes a proprietary map
provider in the event they (Google) changes their billing model, which they
did. That's ok for Google. They can do whatever they like, but Google's api
key and associated online registration became so problematic for me to
figure out, it forced a re-evaluation of mapping tile providers. I'm making
an open-source project and do not want to place a high barrier to entry to
new users with the complex Google maps api key and online registration just
to get Bokeh working.

The OSM and ESRI imagery are high quality, professional alternatives to
Google but are still open source.

Marc

**Comment by austintpg** [comment]
Hi Marc,

It is a good job what you did with this modification, it is working very well for me. I could even make it work inside automatically craeted docker containers, to use it with gitlab. 

Ausitn

**Comment by comperem** [comment]
Hello Austin,

I’m glad to hear of your success. This modification was minor and obvious.
The Bokeh project and organizers have done excellent work over many years.

There are very few open source environments with high quality
documentation, packaging, support, basemaps and a programmable Python loop
*in* *the* *browser*. The periodic callbacks and native html output are
what make bokeh so appealing from my perspective.

Marc

**Comment by bryevdv** [comment]
FYI Bokeh 2.0 with this change was released today

**Comment by austintpg** [comment]
Hi Bryan, that's great! I also want to congratulate for your excellent work, like Marc said before.

**Comment by bryevdv** [comment]
Thanks everyone for the kind words

**Comment by meenurajapandian** [comment]
Hi, I believe the openstreetmap can be used along with bokeh now? But I can't find it in the documentation. Can someone direct me to the right source to use OSM related files with bokeh?

**Comment by bryevdv** [comment]
@meenurajapandian It's the same as all the previous usage. You need to pass a `Vendors` enum, e.g. `Vendors.OSM` (or a string "OSM") to the `get_providers` function. General docs are here:

https://docs.bokeh.org/en/latest/docs/user_guide/geo.html

In the future, please direct support/usage questions to the Discourse: https://discourse.bokeh.org

**Pull request merged**